contribute/architecture/state.md

```diff
@@ -0,0 +1,142 @@
+# State
+
+Grafana uses a database to persist settings between restarts. In fact, if you don't specify one, Grafana creates a [SQLite3](https://www.sqlite.org/) database file on your local disk. This guide explains how to store and retrieve data from the database.
+
+Grafana supports the following databases:
+
+- [MySQL](https://www.mysql.com/)
+- [PostgreSQL](https://www.postgresql.org/)
+- [SQLite3](https://www.sqlite.org/)
+
+Grafana uses the [XORM](https://xorm.io) framework for persisting objects to the database. For more information on how to use XORM, refer to the [documentation](http://gobook.io/read/github.com/go-xorm/manual-en-US/).
+
+[Services](services.md) don't use XORM directly. Instead, services use the _SQL store_, a special type of service that provides an abstraction for the database layer. There are two ways of using the `sqlstore`: using `sqlstore` handlers, and using the `SqlStore` instance.
+
+## `sqlstore` handlers
+
+> **Deprecated:** We are deprecating `sqlstore` handlers in favor of using the `SqlStore`  object directly in each service. Since most services still use the `sqlstore` handlers, we still want to explain how they work.
+
+The `sqlstore` package allows you to register command handlers that either store, or retrieve objects from the database. `sqlstore` handlers are similar to services:
+
+- Services are command handlers that _contain business logic_.
+- `sqlstore` handlers are command handlers that _access the database_.
+
+More importantly, command handlers for services are _methods_, while `sqlstore` handlers are _functions_ that access global variables.
```

Comment:
This sentence is a bit unclear to me. What is the difference between a method and a function? Are stating that functions belong to a struct are methods as opposed to package-wide functions?

Sorry if this a dumb question but I don't really understand what we're trying to convey with this.

Reply:
Go is interesting because it allows you to have both functions and methods, whereas many OOP languages only supports methods. C only supports functions for example. In short:

Method:
```
func (s *MyStruct) MyMethod(s string) error {
}
```

Function:
```
func MyMethod(s *MyStruct, s string) error {
}
```

I personally use functions primarily, and reserve methods for _behavior_ related to the struct.

For a `Server` object, I'd say that `Start` and `Stop` could be considered behavior of a Server, but something like `parseConfigurationFile` is not, as should be a function instead.

Reply:
Apologies as I could have been clearer with the intention of my question. It was oriented towards the intention of the sentence.

Why is it important that we understand this distinction (as it reads "more importantly")? Certainly, it seems like an implementation detail and something that the compiler would be able to tell me when I try to misuse the library.

Reply:
Sorry for dragging this conversation - when something reads "more importantly" my brain immediately triggers into panic mode. I get a sense of urgency that this is a detail that I must absolutely get right, and store it in my head otherwise what I'm building won't work.

From my understanding, it is not the case (and I can certainly be wrong).

Reply:
Oh! I'd be up for changing it. Maybe "Also", or "Additionally"?

Reply:
That would work.

I'd err on the side of removing it altogether as I don't feel it adds anything that hasn't been explained before.